libqtile/drawer.py

```diff
@@ -244,11 +228,74 @@ def __init__(self, qtile, wid, width, height):
         self.clear((0, 0, 1))
 
     def finalize(self):
-        self.qtile.conn.conn.core.FreeGC(self.gc)
-        self.qtile.conn.conn.core.FreePixmap(self.pixmap)
+        del self.gc
+        del self.pixmap
         self.ctx = None
         self.surface = None
 
+    @property
+    def gc(self):
+        if self._gc is None:
+            self._gc = self.qtile.conn.conn.generate_id()
+            self.qtile.conn.conn.core.CreateGC(
+                self._gc,
+                self.wid,
+                xcffib.xproto.GC.Foreground | xcffib.xproto.GC.Background,
+                [
+                    self.qtile.conn.default_screen.black_pixel,
+                    self.qtile.conn.default_screen.white_pixel,
+                ],
+            )
+
+        return self._gc
+
+    @gc.deleter
+    def gc(self):
+        if self._gc is not None:
+            self.qtile.conn.conn.core.FreeGC(self.gc)
+            self._gc = None
+
+    @property
+    def pixmap(self):
+        if self._pixmap is None:
+            self._pixmap = self.qtile.conn.conn.generate_id()
+            self.qtile.conn.conn.core.CreatePixmap(
+                self.qtile.conn.default_screen.root_depth,
+                self._pixmap,
+                self.wid,
+                self.width,
+                self.height
+            )
+            # surface is not guaranteed to be available before pixmap is
+            if self.surface is not None:
+                self.surface.drawable = self._pixmap
+
+        return self._pixmap
+
+    @pixmap.deleter
```

Comment:
The idea itself looks good to me, but can we get rid of some of these properties and just call the actual code we want to call? e.g. here we could just call this:

```
def free_pixmap(self):
    ...
```

and then at user sites we just do ```self.free_pixmap()``` instead of `del self.pixmap`? Feels a little less magical to me that way.

Reply:
Done.